### DISCUSSION TITLE
AutoSquaring - Understanding Dual Axis settings

### DISCUSSION THREAD

**Discussion by Klippert** [grblHAL/core#560] [Q&A] [answered]
Hello,

I've read Phils nice writeup auto squaring https://www.grbl.org/single-post/it-s-hip-to-be-auto-square - and that's all good. I think I understand that correctly (now there's an assumption ;)) 

If i understand it correctly the homing cycle should be something like: (my machine has Y axis with the gantry and two motors.) 
Home Z
Home X
Home Y (Y and B are granged in my case)
- Both Y motors drive towards 0 at the speed set in $25 (Homing search feed rate), until one of the Y endstops is triggered. (I assume that it doesn't care wich one triggers first or if they both trigger?)
- Both Y motors back off a certain distance (set in $27? or am i wrong)
- Both Y motors go towards 0 at the speed set in $24 (Homing Locate feed rate) until one of the Y endstops is triggered.
- Now the Y motor that did not trigger an endstop will move forward with the speed set in $24 (Homing locate feed rate) until it hits the endstop and the gantry is squared. 

Is that right, or am i completely off?
And most of all, how does the dual axis settings work?
What does $347 (Dual axis length fail) mean? I mean as i run integrated servos, if one of them start to lag behind, it will trigger an E-stop. 
And likewise $348 and $349 ?  - i guess if i get to understand one, i'll get to understand the others too :-) 

What happens when i try the autosquaring is it crashes into the end of the machine and gives me a Alarm 15 (Alarm_HomingFailAutoSquaringApproach)

Thanks in advance for any help,

Best regards
Kristoffer

**Comment by terjeio** [comment, marked as answer]
> Is that right, or am i completely off?

You are nearly right. The other motor will continue until the endstop is hit (or fails to do so within the configured limit) as the second step.

> What does $347 (Dual axis length fail) mean?

It is how far the other side will move after the first hit the endstop before an alarm is raised. This to avoid pulling the axis too far out of square. The setting is percent of max travel for the axis. $348 and $349 is used to constrain the result from the value calculated with $347. [The code.](https://github.com/grblHAL/core/blob/1c1519dfb259da485886cf3f883748e662e8f1c8/machine_limits.c#L316-L321).

**Reply by Klippert** [reply]
Hello Terjeio, 

Thanks for the quick answer. 
That clears up how it works. - so a default of 3% seems quite high for my taste as my gantry is only 50cm wide, if i pull it 3% of the travel length (56cm) that would be  1,68cm. -that's definitely going to bend something - i'll set it for 0,5% (which would be about 3mm) 

When testing it out, it all seems to work right. That is, i haven't tried forcing it out of alignment and seeing if it can get "back" to square. I will do that later. 

I also noticed a small curiousity - as i've built GRBLhal for 4 axis, my 5 axis controller uses the B axis for the granged Y-axis. I guess that's just default? - somehow i was expecting the A axis to just move along and be B axis. 
never the less it works. 

I have run into an odd problem though, that when i start the spindle to test it, i get an alarm 1 and everything stops, even if no inputs are triggered. I've built grblhal with pwm spindle, as phils board converts that to 0-10v that controls my VFD. Any clue on what that's all about?

Best
Kristoffer

**Reply by terjeio** [reply]
> I have run into an odd problem though, that when i start the spindle to test it, i get an alarm 1 and everything stops

Does it happen if you disconnect the spindle?

**Reply by Klippert** [reply]
Thats peculiar. If i dont power up The vfd it does not give The error.

let me elaborate my thinking... 

The spindle connection is 0-10v made from a PWM signal from the Teensy. It's using a 12v input power on the board.
It has a connection to a Enable and a Direction as well, which is running to a mosfet board, where it's triggering a 12v supply, as the vfd does not respond to 5v from the teensy breakout board. 
But nothing comes back to the teensy board, there is no modbus, no rev counter, nothing that tells grblhal if the spindle is doing this or that. So how can grblhal pop an error?

And this has all been working for the last year - until I added the clearpath servos...

I'm obviously missing something here... but what?

**Reply by terjeio** [reply]
> But nothing comes back to the teensy board, there is no modbus, no rev counter, nothing that tells grblhal if the spindle is doing this or that. So how can grblhal pop an error?

Likely EMI triggering a limit input. What is the `$LEV` command output pre and post starting the spindle (before/after the alarm)? And if you disable hard limits does the alarm go away?

**Reply by Klippert** [reply]
But I'll have a look. But wouldn't a limit triggered be showing up in the iosender as a red light?

**Reply by Klippert** [reply]
Oh, you're right. it triggers the Y endswitch when i start the spindle. well it doesn't even start, i just makes a sound and then the endstop triggers and everything stops. 
As it's all NC endstops, and the gets pulled down when they trigger, i'm kind of surprised how high the EMI must be. 

Any particular clever ways of getting rid of this problem? I mean everything is already running in shielded cables where the shield is connected to ground. Could that be part of the cause?

/K

**Reply by terjeio** [reply]
> As it's all NC endstops, and the gets pulled down when they trigger

NC (normally closed) switches is typically wired to pull inputs low (to GND) when not triggered.

> Any particular clever ways of getting rid of this problem?

Since there are optocouplers for all inputs it must be a really bad EMI spike, perhaps due to current in a shield?
Is the shields connected at both ends? To the machine chassis in one? And is there a separate ground wire in the cable? If so try connect only one end to ground. Do this with the spindle cable (if shielded) and the Y limit input as a start.
It could also be ground loop currents in the supply side that causes this, perhaps connecting the VFD to a separate outlet will help?

**Reply by Klippert** [reply]
Hello,

Ok that's a fairly complicated problem to solve. 
I located the Y2 endswitch to be the problem. if I disconnect that, everything works. So something must be up specifically with that one. I' guess i'll have to experiment a bit more to  see what i can do. Perhaps i'll just replace that one and see if that changes anything. I'll also check that the wiring and grounding is the same for all endstops. Fingers crossed that it'll sort my EMI issue.

Thannks again for your help this far
Kristoffer

**Reply by Klippert** [reply]
Hmn. This seems to be a more persistent problem. 
I swithced out the endstop for  a different one, and it seemed the problem was gone. 
Now i ran a small test program, and whoopsie, it stops due to a y axis endstop being triggered. At random. 

So i could of course try running the program without the endstop attached, as that seems to be ok, and see if that work. But i'm sort of running out of ideas on how to diagnose and fix this problem.. and bright ideas?

**Reply by terjeio** [reply]
What kind of endstop switch/sensor? How is it wired? Add a photo?
And are you sure it is NC when not triggered?

**Reply by Klippert** [reply]
Hello Terjeio,

This is the endstop: https://www.aliexpress.com/item/1005002632701662.html?spm=a2g0o.order_detail.order_detail_item.3.5620f19cBc8pJE
671WR is the model i'm using. It's a NPN and with "white" wired to +5v it's normally closed.

![image](https://github.com/user-attachments/assets/26dfa776-0724-4f37-8bd7-93975690caba)

Brown and white goes to +5v and black is the signal and blue is to ground.


I believe its a clone of this one: https://www.ia.omron.com/product/item/2223/ with a less chinese spec sheet.

When my Y axis are gantried, is there a way to see which of the Y axis has been triggered with the $LEV ?

Thanks again for any help,
Kristoffer

**Reply by terjeio** [reply]
What is the voltage on the input terminal when triggered/not triggered?
If you replace the endstop sensor with a piece of cable, routed the same way as the sensor, but shorted at the end does it still trigger randomly?

**Reply by Klippert** [reply]
I did a bit of testing.

first off:
Y1 Limit: 0.04v when triggered, and 4,29 when open
Y2 Limit: 0,07v when triggered, and 4,16 when open

For comparison, my X Limit is: 4,86v when triggered and 3,67v when open.

My 5v from the Teensy usb is 4,87, which is a bit less than I expected from a USB.

However i've never seen my x limit trigger the machine to stop, but that does seem not to be normally closed, but rather normally open.

I don't quite know if any conclusions can be made from these measurements alone, but if they can, i'm keen to hear a take on them.

Next up i'll do some "shorting" of the limit switches and rerun som test programs, and see if i can make it fail.

**Reply by terjeio** [reply]
> Y1 Limit: 0.04v when triggered, and 4,29 when open

Again, odd. What is the voltage on the input when the switch is disconnected?
If the voltage is less than 4.29 the the switch is not open collector as per the Omron spec but rather actively driven.
And BTW to me it is NO.

> For comparison, my X Limit is: 4,86v when triggered and 3,67v when open.

And this works?

**Reply by Klippert** [reply]
The X-axis endstop works - though in my little world it shouldn't :)
The EMI stop hasn't managed to trip that one. But i should check if i can change the wiring to make it act as NC. 

Disconnecting the two Y-axis endstops at the box level resulted in 4,09v on both. so it seems a bit odd. 
On the X-axis after disconnecting it, there was exactly the same 4,09v. When connected there was still 3,7v when it was open. 

As a last experiment i tried disconnecting at the board level, so nothing was connected to the control board terminals - which produced the same 4.09v. 

So I guess next up is trying to short Y2 limit after homing, and see if i get the same fail.

**Comment by Klippert** [comment]
I decided to move on to a new discussion topic, as this is no longer about auto squaring. 
New issue is: https://github.com/grblHAL/core/discussions/568
